$MirOS: ports/net/djbdns/patches/patch-src_dnscache-conf_c,v 1.4 2008/10/27 21:40:38 tg Exp $
--- src/dnscache-conf.c.orig	Sun Feb 11 21:11:23 2001
+++ src/dnscache-conf.c	Sat Mar 26 14:32:44 2016
@@ -1,6 +1,7 @@
 #include <unistd.h>
 #include <sys/types.h>
 #include <sys/stat.h>
+#include <stdlib.h>
 #include "hasdevtcp.h"
 #ifdef HASDEVTCP
 #include <sys/mkdev.h>
@@ -37,39 +38,13 @@ const char *myip;
 uint32 seed[32];
 int seedpos = 0;
 
-void seed_adduint32(uint32 u)
-{
-  int i;
-
-  seed[seedpos] += u;
-  if (++seedpos == 32) {
-    for (i = 0;i < 32;++i) {
-      u = ((u ^ seed[i]) + 0x9e3779b9) ^ (u << 7) ^ (u >> 25);
-      seed[i] = u;
-    }
-    seedpos = 0;
-  }
-}
-
 void seed_addtime(void)
 {
-  struct taia t;
-  char tpack[TAIA_PACK];
-  int i;
-
-  taia_now(&t);
-  taia_pack(tpack,&t);
-  for (i = 0;i < TAIA_PACK;++i)
-    seed_adduint32(tpack[i]);
 }
 
 int main(int argc,char **argv)
 {
-  seed_addtime();
-  seed_adduint32(getpid());
-  seed_adduint32(getppid());
-  seed_adduint32(getuid());
-  seed_adduint32(getgid());
+  seed[seedpos++] = arc4random();
 
   user = argv[1];
   if (!user) usage();
@@ -89,13 +64,13 @@ int main(int argc,char **argv)
   if (chdir(auto_home) == -1)
     strerr_die4sys(111,FATAL,"unable to switch to ",auto_home,": ");
 
-  fdrootservers = open_read("/etc/dnsroots.local");
+  fdrootservers = open_read(_SYSCONFDIR "/dnsroots.local");
   if (fdrootservers == -1) {
     if (errno != error_noent)
-      strerr_die2sys(111,FATAL,"unable to open /etc/dnsroots.local: ");
-    fdrootservers = open_read("/etc/dnsroots.global");
+      strerr_die2sys(111,FATAL,"unable to open " _SYSCONFDIR "/dnsroots.local: ");
+    fdrootservers = open_read(_SYSCONFDIR "/dnsroots.global");
     if (fdrootservers == -1)
-      strerr_die2sys(111,FATAL,"unable to open /etc/dnsroots.global: ");
+      strerr_die2sys(111,FATAL,"unable to open " _SYSCONFDIR "/dnsroots.global: ");
   }
 
   init(dir,FATAL);
@@ -114,7 +89,7 @@ int main(int argc,char **argv)
   seed_addtime(); perm(0644);
   seed_addtime(); start("env/IP"); outs(myip); outs("\n"); finish();
   seed_addtime(); perm(0644);
-  seed_addtime(); start("env/IPSEND"); outs("0.0.0.0\n"); finish();
+  seed_addtime(); start("env/IPSEND"); outs("::\n"); finish();
   seed_addtime(); perm(0644);
   seed_addtime(); start("env/CACHESIZE"); outs("1000000\n"); finish();
   seed_addtime(); perm(0644);
@@ -144,6 +119,8 @@ int main(int argc,char **argv)
   seed_addtime(); perm(0644);
   seed_addtime();
 
+  while (seedpos < 32)
+    seed[seedpos++] = arc4random();
   start("seed"); out((char *) seed,128); finish();
   perm(0600);
 
